patches/tModLoader/Terraria/ModLoader/MenuLoader.cs

```diff
@@ -144,11 +144,12 @@ public static class MenuLoader
 		}
 
 		internal static void Unload() {
-			currentMenu = MenutML; // Prevent asset disposed exceptions by disallowing modded menus during the unload process.
-
 			loading = true;
+			// Prevent asset disposed exceptions by disallowing modded menus during the unload process.
 			if (menus.IndexOf(currentMenu) >= DefaultMenuCount) {
-				switchToMenu = MenutML;
+				currentMenu.OnDeselected();
+				currentMenu = MenutML;
+				currentMenu.OnSelected();
```

Comment:
We should only change `currentMenu` on the main thread, that's why `switchToMenu` exists.

I suggest
```cs
switchToMenu = MenutML;
while (currentMenu != MenutML) Thread.Yield();
```

Reply:
Then why is the current behaviour to directly change `currentMenu`?

Reply:
Someone probably added that as a patch that didn't get proper review. selectedMenu is used in the condition